### PR TITLE
tests/core/kernel-and-base-single-reboot-failover, tests/lib/fakestore: verify failover scenario

### PR DESCRIPTION
**Pull request by bboozzoo**
Verify single reboot and failover

### CODE REVIEW

File: tests/core/kernel-and-base-single-reboot-failover/task.yaml [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+        # make sure the system is in stable state, no pending reboots
+        # XXX systemctl exits with non-0 when in degraded state
+        (systemctl is-system-running || true) | MATCH '(running|degraded)'
+
+        # we're expecting the old revisions to be back
+        expecting_kernel="$(cat pc-kernel.rev)"
+        expecting_core="$(cat core.rev)"
+
+        # verify that current points to new revisions
```
Comment (stolowski) [inline tests/core/kernel-and-base-single-reboot-failover/task.yaml:111]: s/new/old/ ?

File: tests/lib/fakestore/cmd/fakestore/cmd_make_refreshable.go [hunk 2]
```diff
@@ -24,12 +24,16 @@ import (
 )
 
 type cmdMakeRefreshable struct {
-	TopDir string `long:"dir" description:"Directory to be used by the store to keep and serve snaps, <dir>/asserts is used for assertions"`
+	TopDir     string `long:"dir" description:"Directory to be used by the store to keep and serve snaps, <dir>/asserts is used for assertions"`
+	SnapBlob   string `long:"snap-blob"`
```
Comment (stolowski) [inline tests/lib/fakestore/cmd/fakestore/cmd_make_refreshable.go:28]: It's a bit confusing how snap-blob is different from snap name, can you add a description?

File: tests/lib/fakestore/cmd/fakestore/cmd_make_refreshable.go [hunk 3]
```diff
@@ -24,12 +24,16 @@ import (
...
+	SnapBlob   string `long:"snap-blob" description:"File or directory with new snap revision contents"`
+	Positional struct {
+		SnapName string
+	} `positional-args:"yes"`
 }
 
 func (x *cmdMakeRefreshable) Execute(args []string) error {
 	// setup fake new revisions of snaps for refresh
-	return refresh.MakeFakeRefreshForSnaps(args, x.TopDir)
+	return refresh.MakeFakeRefreshForSnaps(x.Positional.SnapName, x.TopDir, x.SnapBlob)
```
Comment (stolowski) [inline tests/lib/fakestore/cmd/fakestore/cmd_make_refreshable.go:36]: So we don't support setting multiple snaps at once anymore and instead the tool has to be called for every snap? That's fine, I guess if we shouldn't error out if there are extra args passed (or will it already error out?)

**Review by stolowski — APPROVED** [review]
LGTM, one more question about cmd_make_refreshable.

File: tests/core/kernel-and-base-single-reboot-failover/task.yaml [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        exit
+    fi
+    snap ack "$TESTSLIB/assertions/testrootorg-store.account-key"
+    #shellcheck source=tests/lib/store.sh
+    . "$TESTSLIB"/store.sh
+
+    setup_fake_store "$BLOB_DIR"
+
+    readlink /snap/pc-kernel/current > pc-kernel.rev
+    readlink "/snap/core18/current" > core.rev
```
Comment (mardy) [inline tests/core/kernel-and-base-single-reboot-failover/task.yaml:21]: nitpick^1000: it looks a bit weird that one line has the path quoted, and the other one doesn't.

File: tests/core/kernel-and-base-single-reboot-failover/task.yaml [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        MATCH 'snap_try_kernel=pc-kernel_.*.snap' < boot-vars.dump
+
+        # restore shutdown so that spread can reboot the host
+        tests.cleanup pop
+
+        REBOOT
+    elif [ "$SPREAD_REBOOT" = 1 ]; then
+        change_id="$(cat refresh-change-id)"
+        # XXX: is this sufficiently robust?
+        snap watch "$change_id" || true
```
Comment (mardy) [inline tests/core/kernel-and-base-single-reboot-failover/task.yaml:89]: Is the `|| true` needed here? It seems that `snap watch` returns true if the change has already completed.

File: tests/core/kernel-and-base-single-reboot-failover/task.yaml [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        snap changes | MATCH "$change_id\s+Error"
```
Comment (mardy) [inline tests/core/kernel-and-base-single-reboot-failover/task.yaml:90]: It was not immediately clear to me why this change would have resulted in an error, can you please add a comment for this? Maybe (if my understanding is right):

    # We tried to refresh to a broken kernel snap, so we do expect a failure

**Review by mardy — APPROVED** [review]
LGTM, just a question and a couple of nitpicks :-)

File: tests/core/kernel-and-base-single-reboot-failover/task.yaml (hunk 5 above)
Comment (bboozzoo) [reply tests/core/kernel-and-base-single-reboot-failover/task.yaml:89]: It returns with 0 if the change has succeeded as a whole, but we expect the re-refresh task to have failed, so the whole change will be marked as errored. Probably worth a comment

Comment (bboozzoo) [reply tests/core/kernel-and-base-single-reboot-failover/task.yaml:89]: Err, my bad, was thinking about other scenario. In general yes, snap change will have a non-0 if the change fails (like this one here)

**Pull request merged**